### PR TITLE
Feature/oo 4933/updates for api client

### PR DESCRIPTION
**Pull request by inbalelkabetz**

### CODE REVIEW

File: api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
+package com.takipi.api.client.data.alertssetings;
+
+import com.takipi.api.core.result.intf.ApiResult;
+
+public class DefaultAlertsSettings implements ApiResult
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java]: Why is the word `Default` everywhere?

File: api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package com.takipi.api.client.data.alertssetings;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java]: This is an ApiResult class, should be in `.result`

File: api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+public class DefaultAlertsSettings implements ApiResult
+{
+	public AlertsDefaultEmailSettings email;
+	public AlertsDefaultSlackSettings slack;
+	public AlertsDefaultHipChatSettings hip_chat;
+	public AlertsDefaultPagerDutySettings pager_duty;
+	public AlertsDefaultWebhookSettings webhook;
+	public AlertsDefaultServiceNowSettings service_now;
+	
+	public InitializedFields initializedFields;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java]: This is unrelated to the result in anyway and shouldn't be a member of it

File: api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	
+	public static class AlertsDefaultServiceNowSettings
+	{
+		public String url;
+		public String user_id;
+		public String password;
+		public String table;
+	}
+	
+	public static class InitializedFields
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java]: This (at most) should be an internal class of the `AlertRequest` class, if you really need a class to hold 3 booleans

File: api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsers.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package com.takipi.api.client.data.team;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsers.java]: This is an `ApiResult` class, should be in `.result`

File: api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsers.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+import com.takipi.api.core.result.intf.ApiResult;
+
+import java.util.List;
+
+public class ServiceUsers implements ApiResult
+{
+	public List<TeamMember> team_members;
+	
+	public static class TeamMember
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsers.java]: This is the data class, that should be in `.data`

File: api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsersResponse.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.takipi.api.client.data.team;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsersResponse.java]: This is an `ApiResult` class, should be in `.result`

File: api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsersResponseMessage.java [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package com.takipi.api.client.data.team;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsersResponseMessage.java]: This is an `ApiResult` class, should be in `.result`

File: api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+import com.takipi.api.client.request.ServiceRequest;
+import com.takipi.api.client.result.EmptyResult;
+import com.takipi.api.core.request.intf.ApiPostRequest;
+import com.takipi.common.util.JsonUtil;
+
+import java.util.Map;
+
+public class UpdateAlertsSettingsRequest extends ServiceRequest implements ApiPostRequest<EmptyResult>
+{
+	private final DefaultAlertsSettings defaultAlertsSettings;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java]: This is an `ApiResult` class, and should be used as such.
This class can explicitly hold the actual settings classes as members.

File: api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+		
+		if (defaultAlertsSettings.initializedFields.emailAlertAdditionalEmailsInitialized) {
+			emailsMap.put("alert_additional_emails", Boolean.toString(defaultAlertsSettings.email.alert_additional_emails));
+		}
+		
+		if (defaultAlertsSettings.email.additional_emails_to_alert != null) {
+			emailsMap.put("additional_emails_to_alert", JsonUtil.stringify(defaultAlertsSettings.email.additional_emails_to_alert));
+		}
+		
+		if (emailsMap.values().size() > 0)
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java]: Why check `emailsMap.values().size() > 0` and not `emailsMap.isEmpty()`?

File: api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+			map.put("email", JsonUtil.createSimpleJson(emailsMap));
+		}
+		
+		Map<String, String> slackMap = Maps.newHashMap();
+		
+		if (defaultAlertsSettings.slack.inhook_url != null) {
+			slackMap.put("inhook_url", JsonUtil.stringify(defaultAlertsSettings.slack.inhook_url));
+		}
+		
+		if (slackMap.values().size() > 0)
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java]: Why check `slackMap.values().size() > 0` and not `slackMap.isEmpty()`?

File: api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+			map.put("hip_chat", JsonUtil.createSimpleJson(hipChatMap));
+		}
+		
+		Map<String, String> pagerDutyMap = Maps.newHashMap();
+		
+		if (defaultAlertsSettings.pager_duty.service_integration_key != null) {
+			pagerDutyMap.put("service_integration_key", JsonUtil.stringify(defaultAlertsSettings.pager_duty.service_integration_key));
+		}
+		
+		if (pagerDutyMap.values().size() > 0)
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java]: `isEmpty`

File: api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+		
+		if (defaultAlertsSettings.hip_chat.room != null) {
+			hipChatMap.put("room", JsonUtil.stringify(defaultAlertsSettings.hip_chat.room));
+		}
+		
+		if (defaultAlertsSettings.hip_chat.url != null) {
+			hipChatMap.put("url", JsonUtil.stringify(defaultAlertsSettings.hip_chat.url));
+		}
+		
+		if (hipChatMap.values().size() > 0)
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java]: `isEmpty`

File: api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+			map.put("pager_duty", JsonUtil.createSimpleJson(pagerDutyMap));
+		}
+		
+		Map<String, String> webhookMap = Maps.newHashMap();
+		
+		if (defaultAlertsSettings.webhook.webhook_url != null) {
+			webhookMap.put("webhook_url", JsonUtil.stringify(defaultAlertsSettings.webhook.webhook_url));
+		}
+		
+		if (webhookMap.values().size() > 0)
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java]: `isEmpty`

File: api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+			{
+				throw new IllegalArgumentException("No webhook URL supplied for webhook request");
+			}
+			
+			if ((this.defaultAlertsSettings.service_now.url != null) ||
+				(this.defaultAlertsSettings.service_now.user_id != null) ||
+				(this.defaultAlertsSettings.service_now.password != null) ||
+				(this.defaultAlertsSettings.service_now.table != null))
+			{
+				if ((Strings.isNullOrEmpty(this.defaultAlertsSettings.service_now.url)) ||
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/alertssettings/UpdateAlertsSettingsRequest.java]: This check basically says "if any of those params isn't null, all of them has to be non-null-non-empty"

However, in the request `postData` you explicitly only add them one-by-one.
This tells me that the API endpoint allows you to just change one of the parameters.

If the API endpoint does allow that, then the check here is incorrect, as it prohibits something the API endpoint allows.
If the API endpoint doesn't allow that, the then building inside `postData` is doing a lot of unnecessary work.

This holds true for a lot of the settings here.

File: api-client/src/main/java/com/takipi/api/client/request/team/ChangeTeamMembersRolesRequest.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+			
+			return this;
+		}
+		
+		@Override
+		protected void validate()
+		{
+			super.validate();
+			
+			if (this.members.size() == 0)
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/ChangeTeamMembersRolesRequest.java]: `isEmpty`

File: api-client/src/main/java/com/takipi/api/client/request/team/GetTeamMembersRequest.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
+package com.takipi.api.client.request.team;
+
+import com.takipi.api.client.data.team.ServiceUsers;
+import com.takipi.api.client.request.ServiceRequest;
+import com.takipi.api.core.request.intf.ApiGetRequest;
+
+public class GetTeamMembersRequest extends ServiceRequest implements ApiGetRequest<ServiceUsers>
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/GetTeamMembersRequest.java]: Why isn't this `ServiceUsersResponse`?

File: api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsersResponse.java [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package com.takipi.api.client.data.team;
+
+import com.takipi.api.core.result.intf.ApiResult;
+
+import java.util.List;
+
+public class ServiceUsersResponse implements ApiResult
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/data/team/ServiceUsersResponse.java]: `ServiceUsersResult`

File: api-client/src/main/java/com/takipi/api/client/request/team/GetTeamMembersRequest.java [hunk 19]
```diff
@@ -0,0 +1,45 @@
+package com.takipi.api.client.request.team;
+
+import com.takipi.api.client.data.team.ServiceUsers;
+import com.takipi.api.client.request.ServiceRequest;
+import com.takipi.api.core.request.intf.ApiGetRequest;
+
+public class GetTeamMembersRequest extends ServiceRequest implements ApiGetRequest<ServiceUsers>
+{
+protected GetTeamMembersRequest (String serviceId) {
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/GetTeamMembersRequest.java]: The formatting in this class makes no sense

File: api-client/src/main/java/com/takipi/api/client/request/team/DeleteTeamMembersRequest.java [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+import com.takipi.api.client.data.team.ServiceUsers.TeamMember;
+import com.takipi.api.client.data.team.ServiceUsersResponseMessage;
+import com.takipi.api.client.request.ServiceRequest;
+import com.takipi.api.core.request.intf.ApiDeleteRequest;
+
+import java.util.List;
+
+public class DeleteTeamMembersRequest extends ServiceRequest implements ApiDeleteRequest<ServiceUsersResponseMessage>
+{
+	private final ServiceUsers serviceUsers;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/DeleteTeamMembersRequest.java]: This is an unnecessary wrapper for a collection of `TeamMember`

File: api-client/src/main/java/com/takipi/api/client/request/team/ChangeTeamMembersRolesRequest.java [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+import com.takipi.api.client.data.team.ServiceUsers.TeamMember;
+import com.takipi.api.client.data.team.ServiceUsersResponseMessage;
+import com.takipi.api.client.request.ServiceRequest;
+import com.takipi.api.core.request.intf.ApiPostRequest;
+
+import java.util.List;
+
+public class ChangeTeamMembersRolesRequest extends ServiceRequest implements ApiPostRequest<ServiceUsersResponseMessage>
+{
+	private final ServiceUsers serviceUsers;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/ChangeTeamMembersRolesRequest.java]: This is an unnecessary wrapper for a collection of `TeamMember`

File: api-client/src/main/java/com/takipi/api/client/request/team/ChangeTeamMembersRolesRequest.java [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+	public String urlPath() {
+		return baseUrlPath() + "/team/change-role";
+	}
+	
+	public static Builder newBuilder() {
+		return new Builder();
+	}
+	
+	public static class Builder extends ServiceRequest.Builder {
+		private List<TeamMember> members = Lists.newArrayList();
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/ChangeTeamMembersRolesRequest.java]: Unless we allow a user to have multiple roles, this should be map.
Also, please initialize members in the c-tor

File: api-client/src/main/java/com/takipi/api/client/request/team/ChangeTeamMembersRolesRequest.java [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+			
+			for (TeamMember teamMember : this.members)
+			{
+				if ((Strings.isNullOrEmpty(teamMember.email)) ||
+					(Strings.isNullOrEmpty(teamMember.role)))
+				{
+					throw new IllegalArgumentException("User email or role cannot be empty");
+				}
+				
+				if (!((teamMember.role.equalsIgnoreCase("Owner")) ||
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/ChangeTeamMembersRolesRequest.java]: If we only allow explicit roles, this should be an enum (see `ValidationUtil`)

File: api-client/src/main/java/com/takipi/api/client/request/team/DeleteTeamMembersRequest.java [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+	public Class<ServiceUsersResponseMessage> resultClass() {
+		return ServiceUsersResponseMessage.class;
+	}
+	
+	public static Builder newBuilder() {
+		return new Builder();
+	}
+	
+	public static class Builder extends ServiceRequest.Builder {
+		private List<String> usersToRemove = Lists.newArrayList();
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/DeleteTeamMembersRequest.java]: This should be a `Set`
Also, initialize members in the c-tor

File: api-client/src/main/java/com/takipi/api/client/request/team/DeleteTeamMembersRequest.java [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+		}
+		
+		@Override
+		public Builder setServiceId(String serviceId) {
+			super.setServiceId(serviceId);
+			
+			return this;
+		}
+		
+		public DeleteTeamMembersRequest.Builder addTeamMemberToDelete(String email) {
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/DeleteTeamMembersRequest.java]: No need to explicitly say which builder this returns.

File: api-client/src/main/java/com/takipi/api/client/request/team/DeleteTeamMembersRequest.java [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+			
+			return this;
+		}
+		
+		@Override
+		protected void validate()
+		{
+			super.validate();
+			
+			if (this.usersToRemove.size() == 0)
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/DeleteTeamMembersRequest.java]: `isEmpty`

File: api-client/src/main/java/com/takipi/api/client/request/team/UpdateTeamMembersRequest.java [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+import com.takipi.api.client.data.team.ServiceUsers.TeamMember;
+import com.takipi.api.client.data.team.ServiceUsersResponse;
+import com.takipi.api.client.request.ServiceRequest;
+import com.takipi.api.core.request.intf.ApiPostRequest;
+
+import java.util.List;
+
+public class UpdateTeamMembersRequest extends ServiceRequest implements ApiPostRequest<ServiceUsersResponse>
+{
+	private final ServiceUsers serviceUsers;
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/UpdateTeamMembersRequest.java]: This is an unnecessary wrapper for a collection of `TeamMember`

File: api-client/src/main/java/com/takipi/api/client/request/team/UpdateTeamMembersRequest.java [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+import com.google.gson.Gson;
+import com.takipi.api.client.data.team.ServiceUsers;
+import com.takipi.api.client.data.team.ServiceUsers.TeamMember;
+import com.takipi.api.client.data.team.ServiceUsersResponse;
+import com.takipi.api.client.request.ServiceRequest;
+import com.takipi.api.core.request.intf.ApiPostRequest;
+
+import java.util.List;
+
+public class UpdateTeamMembersRequest extends ServiceRequest implements ApiPostRequest<ServiceUsersResponse>
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/UpdateTeamMembersRequest.java]: If this request simply adds members, and doesn't actually let you update any new information, I believe `AddTeamMembersRequest` is a clearer name

File: api-client/src/main/java/com/takipi/api/client/request/team/UpdateTeamMembersRequest.java [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+	public String urlPath() {
+		return baseUrlPath() + "/team";
+	}
+	
+	public static Builder newBuilder() {
+		return new Builder();
+	}
+	
+	public static class Builder extends ServiceRequest.Builder {
+		private List<String> emailsToAdd = Lists.newArrayList();
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/UpdateTeamMembersRequest.java]: This should be a `Set`
Also, initialize members in the c-tor

File: api-client/src/main/java/com/takipi/api/client/request/team/UpdateTeamMembersRequest.java [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+			
+			return this;
+		}
+		
+		@Override
+		protected void validate()
+		{
+			super.validate();
+			
+			if (this.emailsToAdd.size() == 0)
```
Comment (dor-levi) [inline api-client/src/main/java/com/takipi/api/client/request/team/UpdateTeamMembersRequest.java]: `isEmpty`

File: api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java (hunk 1 above)
Comment (inbalelkabetz) [reply api-client/src/main/java/com/takipi/api/client/data/alertssetings/DefaultAlertsSettings.java]: it's the default settings for alerts.

File: api-client/src/main/java/com/takipi/api/client/request/team/GetTeamMembersRequest.java (hunk 17 above)
Comment (inbalelkabetz) [reply api-client/src/main/java/com/takipi/api/client/request/team/GetTeamMembersRequest.java]: different field names.

Comment (dor-levi) [reply api-client/src/main/java/com/takipi/api/client/request/team/GetTeamMembersRequest.java]: In that case I'd rename `ServiceUsers` and `ServiceUsersResponse` to `TeamMembersResult` and `NewTeamMembersResult` so it's more clear which is which.
